02-peer-protocol.md

```diff
@@ -49,11 +68,468 @@ transaction is confirmed are also not persistent - until you know the script
+        |       |<-(2)- tx_complete ----|       |
+        |       |--(3)- tx_add_input -->|       |
+        |   A   |<-(4)- tx_complete ----|   B   |
+        |       |--(5)- tx_add_output ->|       |
+        |       |<-(6)- tx_complete ----|       |
+        |       |--(7)- tx_complete --->|       |
+        +-------+                       +-------+
+
+#### *initiator* and *non-initiator*
+
+A the *initiator* contributes 2 inputs and an output that they
+then remove.  B, the *non-initiator*, contributes 1 input and an output,
+but waits until A adds a second input before contributing.
+
+Note that if A does not send a second input, the negotiation will end without
+B's contributions.
+
+        +-------+                         +-------+
+        |       |--(1)- tx_add_input ---->|       |
+        |       |<-(2)- tx_complete ------|       |
+        |       |--(3)- tx_add_output --->|       |
+        |       |<-(4)- tx_complete ------|       |
+        |       |--(5)- tx_add_input ---->|       |
+        |   A   |<-(6)- tx_add_input -----|   B   |
+        |       |--(7)- tx_remove_output >|       |
+        |       |<-(8)- tx_add_output ----|       |
+        |       |--(9)- tx_complete ----->|       |
+        |       |<-(10) tx_complete ------|       |
+        +-------+                         +-------+
+
+### The `tx_add_input` Message
+
+This message contains a transaction input.
+
+1. type: 66 (`tx_add_input`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+    * [`u16`:`prevtx_len`]
+    * [`prevtx_len*byte`:`prevtx`]
+    * [`u32`:`prevtx_vout`]
+    * [`u32`:`sequence`]
+
+#### Requirements
+
+The sending node:
+  - MUST add all sent inputs to the transaction
+  - MUST use a unique `serial_id` for each input currently added to the
+    transaction
+  - MUST NOT re-transmit inputs it has received from the peer
+  - if is the *initiator*:
+    - MUST send even `serial_id`s
+  - if is the *non-initiator*:
+    - MUST send odd `serial_id`s
+
+The receiving node:
+  - MUST add all received inputs to the transaction
+  - MUST fail the negotiation if:
+    - the `prevtx` and `prevtx_vout` are identical to a previously added
+      (and not removed) input's
+    - `prevtx` is not a valid transaction
+    - `prevtx_vout` is greater or equal to the number of outputs on `prevtx`
+    - the `scriptPubKey` of the `prevtx_vout` output of `prevtx` is not exactly a 1-byte push opcode (for the numeric values `0` to `16`) followed by a data push between 2 and 40 bytes
+    - the `serial_id` is already included in the transaction
+    - the `serial_id` has the wrong parity
+    - if has received 4096 `tx_add_input` messages during this negotiation
+
+#### Rationale
+
+Each node must know the set of the transaction inputs. The *non-initiator*
+MAY omit this message.
+
+`serial_id` is a randomly chosen number which uniquely identifies this input.
+Inputs in the constructed transaction are sorted by `serial_id`.
```

Comment:
Why not use BIP69 ordering?

Using this random ordering may become a marker to identify lightning channels on chain.

Reply:
The reason to use that ordering instead of BIP 69 is to potentially allow usage of `sighash_single` which requires participants to control the order of inputs/outputs.

It's not trivial to effectively control this though: if you use serial IDs 0 and 2 for and input and output that you want to sign using `sighash_single`, there is a risk that your peer will add an output with serial ID 1, which messes that up. You may be able to rectify this using `tx_remove_input` and `tx_remove_output` to insert them at a different place, but that won't work in every case. But in some higher-level protocols that specify how serial IDs must be assigned to allow exotic sighash constructions, this would come in handy and would be impossible with BIP 69. Does that seem like a reasonable trade-off?

Reply:
It's hard for me to judge -- I'm not familiar with all the applications of the `interactive-tx` protocol.

At least for dual funding, I don't see a reason to use `sighash_single`.

Reply:
The only application i can think of is attaching htlc 1st level txs

Reply:
My personal preference is to favor flexibility for protocols that will build on top of `interactive-tx`, because we don't want to restrict what can be built. We can always later add an optional tlv to request using a different ordering (e.g. bip 69) if we feel it's important.

Reply:
> At least for dual funding, I don't see a reason to use sighash_single.

But a dual funding transaction can be batched with another type of transaction (splicing, coinjoin, etc) and you don't even have to know about it (which is great). We don't want to make that impossible in my opinion.